res_syntax/src/reactjs_jsx_v4.ml

```diff
@@ -287,21 +287,39 @@ let makePropsTypeParams ?(stripExplicitOption = false)
          else if isOptional && stripExplicitOption then stripOption interiorType
          else Some interiorType)
 
-let makeLabelDecls ~loc namedTypeList =
-  namedTypeList
-  |> List.map (fun (isOptional, label, attrs, interiorType) ->
-         if label = "key" then
-           Type.field ~loc ~attrs:(optionalAttrs @ attrs) {txt = label; loc}
-             interiorType
-         else if isOptional then
-           Type.field ~loc ~attrs:(optionalAttrs @ attrs) {txt = label; loc}
-             (Typ.var @@ safeTypeFromValue @@ Labelled label)
-         else
-           Type.field ~loc ~attrs {txt = label; loc}
-             (Typ.var @@ safeTypeFromValue @@ Labelled label))
+let makeLabelDecls namedTypeList =
+  let rec mem_fst x = function
+    | [] -> false
+    | (a, _) :: l -> compare a (fst x) = 0 || mem_fst x l
+  in
+  let rec duplicated = function
+    | [] -> None
+    | hd :: tl -> if mem_fst hd tl then Some hd else duplicated tl
+  in
+  let duplicatedProp =
+    (* Find the duplicated prop from the back *)
+    namedTypeList |> List.rev
```

Comment:
Now we always have to do an additional `List.rev` and `List.map` even if there are no duplicated labels at all, in which case we have another `List.map` below.

Could this be done in a more efficient way, avoiding additional List allocations in the standard case (no duplicate props)?
Maybe it would be better

Reply:
Can you elavorate your thought? Do you mean that finding the duplicated prop without `List.rev` first and then `List.rev` if there is duplicated one?

Reply:
Maybe this?
1. check if there is duplicated without `List.map` or `List.exists`
2. if there is, then `List.map` and `List.rev`

Reply:
Yes, something like that, it would be ideal if we could perform the check without creating any additional lists.

Reply:
> Yes, something like that, it would be ideal if we could perform the check without creating any additional lists.

Agree! I'll work on it.

Reply:
Sorry for being late, I'm stuck with tasks from company 🥲 

How does it look? https://github.com/rescript-lang/rescript-compiler/pull/5937/commits/10421cd47c0e5c93ce1375b924e81c717ba10474 Looking a bit verbose, but I think it is avoiding the list relocation before checking for duplicated props.

Reply:
Looks good to me! Thanks a lot!